GH-30302: [C++][Parquet] Preserve the bitwidth of integer dictionary indices on round-trip to Parquet

lesterfan opened the pull request:
### Rationale for this change
This fixes the bug reported in Github issue https://github.com/apache/arrow/issues/30302.

### What changes are included in this PR?
This PR does the following:
1. Replaces the hardcoded `int32` inferred dictionary index type with the faithful index type stored in the parquet file metadata.
1. Replaces a `View` call with a more permissive `ViewOrCastChunkedArray` in `TransferDictionary` where the built dictionary is transferred to the caller. This function does an explicit cast if the `View` fails, allowing e.g. `dictionary<values=string, indices=int32, ordered=0>` to be interpreted as `dictionary<values=string, indices=int8, ordered=0>` even though the layouts aren't equal.
1. Adds a unit test which passes on this branch and fails on main (due to the schema types `dictionary<values=string, indices=int8, ordered=0>`, `dictionary<values=string, indices=int32, ordered=0>` not being equal).

Note that there is a performance overhead when casting. I did the following experiment to measure this.

With the code:
```
#include <stdio.h>

#include <filesystem>
#include <fstream>
#include <iostream>
#include <string>

#include "arrow/api.h"
#include "arrow/compute/api.h"
#include "arrow/io/api.h"
#include <arrow/api.h>
#include <arrow/io/file.h>
#include <parquet/arrow/reader.h>
#include <parquet/arrow/writer.h>

arrow::Status arrow_test() {
    arrow::MemoryPool* pool = arrow::default_memory_pool();

    std::shared_ptr<arrow::Array> values;
    arrow::DictionaryBuilder<arrow::StringType> dict_builder(pool);
    for (int i = 0; i < 100'000'000; ++i) {
        ARROW_RETURN_NOT_OK(dict_builder.Append("abc"));
        ARROW_RETURN_NOT_OK(dict_builder.Append("def"));
        ARROW_RETURN_NOT_OK(dict_builder.Append("ghi"));
    }
    ARROW_RETURN_NOT_OK(dict_builder.Finish(&values));

    auto dict_type = dict_builder.type();
    auto schema = arrow::schema({arrow::field("x", dict_type)});

    auto table = arrow::Table::Make(schema, {values});

    const char* filepath = "/not/a/real/filepath"; // Replace with a real filepath
    std::shared_ptr<arrow::io::FileOutputStream> outfile;
    arrow::io::FileOutputStream::Open(filepath).ValueOrDie().swap(outfile);

    auto writer_props = parquet::WriterProperties::Builder().build();
    // This is important; if this is not set, the writer won't add the
    // "ARROW:schema": <base 64 encoded schema> metadata to the parquet file,
    // and the reader will read the column as a string rather than a dictionary.
    auto arrow_writer_props =
            parquet::ArrowWriterProperties::Builder().store_schema()->build();

    if (auto write_table_status = parquet::arrow::WriteTable(
                *table, pool, outfile, 1024, writer_props, arrow_writer_props);
        !write_table_status.ok()) {
        printf("Error writing table to file: %s\n",
               write_table_status.ToString().c_str());
        return write_table_status;
    }

    std::shared_ptr<arrow::io::ReadableFile> infile;
    arrow::io::ReadableFile::Open(filepath).ValueOrDie().swap(infile);
    std::unique_ptr<parquet::arrow::FileReader> reader;
    ARROW_ASSIGN_OR_RAISE(reader, parquet::arrow::OpenFile(infile, pool));

    std::shared_ptr<arrow::Table> table_round_trip;
    ARROW_RETURN_NOT_OK(reader->ReadTable(&table_round_trip));

    printf("table: %s\n", table->ToString().c_str());
    printf("table_round_trip: %s\n", table_round_trip->ToString().c_str());

    return arrow::Status::OK();
}

int main() {
    auto status = arrow_test();
    if (!status.ok()) {
        printf("Not ok!, status: %s\n", status.ToString().c_str());
        return 1;
    }
    printf("ok\n");
}
```
I wrote a parquet file with 300 million rows but only three unique values (which fit into an `int8_t`). I then commented out the code which writes the parquet file to purely measure the read performance. Here are the results:
```
(without the change locally, 300M rows):
./reproduce_experiment  125.67s user 1.13s system 98% cpu 2:08.88 total
(with the change locally, 300M rows):
./reproduce_experiment  137.25s user 1.42s system 99% cpu 2:19.60 total
```
The performance impact is minimal and only in the case where a cast is necessary (i.e. this impact would not exist if the written type had an `int32` index). I would argue this is preferable to the status quo, where the type returned is incorrect (i.e. not faithful to the original type written).

A potential implementation which wouldn't involve this casting would involve templating the `::arrow::BinaryDictionary32Builder` [here](https://github.com/apache/arrow/blob/784aa6faf69f5cf135e09976a281dea9ebf58166/cpp/src/parquet/column_reader.cc#L2176) by the index type (instead of a hard-coded `int32` type) so that the correct bit-width is used while reading the file. This would be a more involved change.

### Are these changes tested?
Yes, though the test is fairly basic; let me know if you guys have any suggestions on how to improve this!

### Are there any user-facing changes?
Yes, this PR implements a bug fix to a user facing API.

* GitHub Issue: #30302

pitrou commented on cpp/src/parquet/arrow/reader_internal.cc:
```diff
@@ -256,6 +256,23 @@ Status ByteArrayStatisticsAsScalars(const Statistics& statistics,
+    return view_result.ValueOrDie();
+  } else {
+    ::arrow::ArrayVector casted_chunks;
+    for (const auto& chunk : array->chunks()) {
+      ARROW_ASSIGN_OR_RAISE(auto cast_chunk,
+                            ::arrow::compute::Cast(chunk, logical_value_type));
+      casted_chunks.push_back(cast_chunk.make_array());
+    }
```
Two comments here:
1. There is a `Cast` function overload that accepts a `Datum`, why not use that instead of casting each chunk individually?
2. We should pass the right `MemoryPool` when casting.

lesterfan replied:
This makes sense; I didn't know about (1). Maybe a dumb question for (2), but how should I determine which `MemoryPool` to use?

pitrou commented on cpp/src/parquet/arrow/arrow_reader_writer_test.cc:
```diff
@@ -3336,6 +3336,38 @@ TEST(TestArrowReadWrite, NonUniqueDictionaryValues) {
   }
 }
 
+TEST(TestArrowReadWrite, DictionaryIndexBidWidthsArePreservedOnRoundTrip) {
```
Suggestion: 1) fix typo 2) make test name slightly shorter :)
```suggestion
TEST(TestArrowReadWrite, DictionaryIndexBitwidthRoundtrip) {
```

pitrou commented on cpp/src/parquet/arrow/arrow_reader_writer_test.cc:
```diff
@@ -3336,6 +3336,38 @@ TEST(TestArrowReadWrite, NonUniqueDictionaryValues) {
+  ArrayFromVector<::arrow::Int8Type, int8_t>({2, 0, 1, 0, 2}, &f1_values);
+  ::arrow::ArrayVector dict_arrays = {
+      std::make_shared<::arrow::DictionaryArray>(dict_type, f0_values, dict_values),
+      std::make_shared<::arrow::DictionaryArray>(dict_type, f1_values, dict_values)};
```
Ok, we have a facility named `DictArrayFromJSON` (you can grep for it) that will make constructing the dictionary arrays much easier.

pitrou reviewed (changes requested):
Thanks for investigating this and proposing a fix @lesterfan !
This looks good on the principle, here are a couple questions and suggestions.

pitrou commented on cpp/src/parquet/arrow/reader_internal.cc:
```diff
@@ -256,6 +256,23 @@ Status ByteArrayStatisticsAsScalars(const Statistics& statistics,
+Result<std::shared_ptr<ChunkedArray>> ViewOrCastChunkedArray(
+    const std::shared_ptr<ChunkedArray>& array,
+    const std::shared_ptr<DataType>& logical_value_type) {
+  auto view_result = array->View(logical_value_type);
```
Oh, I see, thanks.

lesterfan commented:
Thank you for the quick review @pitrou! I just updated my PR; let me know if you see anything else! I'm also going to re-run the experiment described above now that we're doing the cast in one shot and will edit the results in the description accordingly.

lesterfan commented:
> I'm also going to re-run the experiment described above now that we're doing the cast in one shot and will edit the results in the description accordingly.

This is done. I didn't find anything too notable here. 

Before I updated the PR:
```
(without the change locally, 300M rows):
./reproduce_experiment  125.04s user 1.01s system 98% cpu 2:07.80 total
(with the change locally, 300M rows):
./reproduce_experiment  139.79s user 0.98s system 99% cpu 2:20.95 total
```
After I updated the PR:
```
(without the change locally, 300M rows):
./reproduce_experiment  125.67s user 1.13s system 98% cpu 2:08.88 total
(with the change locally, 300M rows):
./reproduce_experiment  137.25s user 1.42s system 99% cpu 2:19.60 total
```
I updated the top level description with the new results.

pitrou commented on cpp/src/parquet/arrow/arrow_reader_writer_test.cc:
```diff
@@ -3336,6 +3336,26 @@ TEST(TestArrowReadWrite, NonUniqueDictionaryValues) {
 }
 
+TEST(TestArrowReadWrite, DictionaryIndexBitwidthRoundtrip) {
+  auto dict_type = ::arrow::dictionary(::arrow::int8(), ::arrow::utf8());
```
Let's perhaps be nifty and check all possible index types? Something like:
```suggestion
  for (auto index_type : {int8(), int16(), int32(), int64()}) {
    ARROW_SCOPED_TRACE("index_type = " , *index_type);
    auto dict_type = ::arrow::dictionary(index_type, ::arrow::utf8());
    ...
```

wgtmac commented on cpp/src/parquet/arrow/schema.cc:
```diff
@@ -989,8 +989,8 @@ Result<bool> ApplyOriginalStorageMetadata(const Field& origin_field,
         checked_cast<const ::arrow::DictionaryType&>(*origin_type);
-    inferred->field = inferred->field->WithType(
-        ::arrow::dictionary(::arrow::int32(), inferred_type, dict_origin_type.ordered()));
+    inferred->field = inferred->field->WithType(::arrow::dictionary(
```
I think this fix makes sense, though it might introduce a benign breaking change to the downstream users.

pitrou reviewed (commented):
Thank you, just one more suggestion.

lesterfan commented:
I think I went through the remainder of the comments. Let me know if I missed anything here!

mapleFU commented on cpp/src/parquet/arrow/reader_internal.cc:
```diff
@@ -547,7 +563,7 @@ Status TransferBinary(RecordReader* reader, MemoryPool* pool,
                       std::shared_ptr<ChunkedArray>* out) {
   if (reader->read_dictionary()) {
```
So still when read_dictionary:
1. If writer set dictionary, it uses writers type
2. Otherwise, use int32 dict?

lesterfan replied:
That is indeed the intention here. In the case where the writer doesn't set the dictionary type, `TransferBinary` will call `TransferDictionary` [here](https://github.com/apache/arrow/blob/d88ef57a737b246c93e07ff98c19ff488bca1c59/cpp/src/parquet/arrow/reader_internal.cc#L549) which still hardcodes the `int32` type.

mapleFU reviewed (commented):
General LGTM

pitrou commented:
@lesterfan I think you want to rebase/merge on git main so as to fix the CI failures.

pitrou commented on r/tests/testthat/test-dataset.R:
```diff
@@ -915,7 +915,7 @@ test_that("Dataset and query print methods", {
       "lgl: bool",
       "chr: string",
-      "fct: dictionary<values=string, indices=int32>",
+      "fct: dictionary<values=string, indices=int8>",
```
Ok, we need to make this test depend on the version of Arrow. Hopefully someone can help me :) @assignUser

assignUser replied:
We have helper functions that skip a test based on the R package version. But I don't think we have a function that skips tests based on libarrow version. But you can use `arrow::arrow_info()$build_info$cpp_version` to get the version the package is build against and add a function to `helper-skip.R`

assignUser replied:
or not skip but change the expected output

pitrou replied:
Given that I'm not able to run the Docker image locally (it fails building) and I'm unlikely to succeed fixing the test without at least a couple attempts, would you perhaps be able to push the required changes to this PR?

thisisnic replied:
Apologies, travelling at the moment, and can take a look tomorrow, but how I'll do it if nobody gets there first; if it's not too prevalent across the tests, something like this.

Add this to `r/tests/testthat/helper-arrow.R`:

```
arrow_cpp_version_at_least <- function(version) {
  current_version <- arrow::arrow_info()$build_info$cpp_version
  numeric_version(current_version) >= numeric_version(version)
}
```

Replace the above test with something like:

```
test_that("Dataset and query print methods", {
  ds <- open_dataset(hive_dir)
  
  dict_index_type <- ifelse(arrow_cpp_version_at_least("19.0.1"), "int8", "int32")

  expect_output(
    print(ds),
    paste(
      "FileSystemDataset with 2 Parquet files",
      "8 columns",
      "int: int32",
      "dbl: double",
      "lgl: bool",
      "chr: string",
      paste("fct: dictionary<values=string, indices=", dict_index_type, ">", sep = ""),
      "ts: timestamp[us, tz=UTC]",
      "group: int32",
      "other: string",
      sep = "\n"
    ),
    fixed = TRUE
  )
....
```

amoeba replied:
I pushed up a slightly tweaked version of the above in https://github.com/apache/arrow/pull/45685/commits/89a424d6235441687abd9e65938cc9a710eece72 and it passes for me against dev and from 19.0.1.

thisisnic replied:
Thanks @amoeba!
